### ISSUE TITLE
refetchQueries not working when using string array after mutation

### ISSUE THREAD

**Issue by reinvanimschoot** [apollographql/apollo-client#5419]
**Intended outcome:**
After a create mutation, I want to refetch the list query to have the newly added item be displayed in the list, making use of the variables which the list query is run with previously.

As per the documentation, this should be doable:

> Please note that if you call refetchQueries with an array of strings, then Apollo Client will look for any previously called queries that have the same names as the provided strings. It will then refetch those queries with their current variables.

**Actual outcome:**
When using the operation name as string in the refetchQueries array, no query is refetched, nothing is updated and no network activity is visible.

**Versions**

```
  System:
    OS: macOS Mojave 10.14.6
  Binaries:
    Node: 10.16.3 - ~/.nvm/versions/node/v10.16.3/bin/node
    Yarn: 1.19.0 - /usr/local/bin/yarn
    npm: 6.9.0 - ~/.nvm/versions/node/v10.16.3/bin/npm
  Browsers:
    Chrome: 77.0.3865.90
    Safari: 13.0.1
  npmPackages:
    @apollo/react-hooks: ^3.1.0 => 3.1.1
    apollo-cache-inmemory: ^1.6.3 => 1.6.3
    apollo-cache-persist: ^0.1.1 => 0.1.1
    apollo-client: ^2.6.4 => 2.6.4
    apollo-link: ^1.2.13 => 1.2.13
    apollo-link-context: ^1.0.19 => 1.0.19
    apollo-link-error: ^1.1.12 => 1.1.12
    apollo-link-http: ^1.5.16 => 1.5.16
    apollo-link-logger: ^1.2.3 => 1.2.3
    react-apollo: ^3.1.2 => 3.1.2
```

**Comment by jinshin1013** [comment]
I had a similar problem but was resolved by providing the exact object used to fetch the query which may include any variables, query options and etc.

**Comment by reinvanimschoot** [comment]
The whole purpose of using a string is that you should NOT have to provide the exact object as Apollo should automatically refetch the query with the current query variables.

**Comment by jinshin1013** [comment]
Yeah I totally agree with you there man it sure is a bug or documentation is wrong. Didn't mean to sound like it ain't a bug. It's very inconvenient having to provide the exact same object especially multiple mutations in multiple places refetches the query.

**Comment by tafelito** [comment]
I have a similar issue. The weird thing is that when I call a mutation where it deletes a row, and then I call the refetchQueries with a string, it actually works. But after I insert a new row, nothing happens when I do the refetch. It makes no sense because the delete is still a mutation, so I'm not sure what the difference could be. Both mutations return the same values. I'm looking at the network requests, and I see that after the insert there is no request, but after the delete, there is a new request that actually do re fetch

**Comment by reinvanimschoot** [comment]
I have witnessed exactly the same behaviour! It works fine when I refetch after I delete an item but not when I create one.

**Comment by mnesarco** [comment]
Hi Friends, I have the same problem: refetchQueries works with update mutations, but nothing happens on insert mutations. Is there any known workaround?

**Comment by kirkchris** [comment]
Impacting us here also... 😫

**Comment by seanraad123** [comment]
Also running into this issue

**Comment by ScottAgirs** [comment]
Same here folks

**Comment by onderonur** [comment]
I came accros this issue.
I create a new item from a modal, while the listing page is in the background.
After the mutation, I redirect to the detail page of the new item.
Also, when the mutation is completed, I can see the listing query is refetched (in browser dev-tools).
But the listing query result in the cache not getting updated.

If I cancel the redirect, which comes right after the mutation, there is no problem. After the query is refetched, cache gets updated.
If I keep the redirect, but make `awaitRefetchQueries: true` in the mutation options, there is no problem again.

I think, when the page that observes the query gets unmounted, even if the network request gets completed, cache can not be updated.

A cool way could be invalidating a query even if it wasn't observed by any active component.
Using this way, one could just tell the cache that the query result is now invalidated. And the first call to that query would be done by a network request. So, user creates a new record. Get redirected to the detail page of the new item. User may go to other pages, it doesn't matter. But when we come to the listing page again, we would trigger a network request, instead of showing the stale data in the cache.

This is just a simple idea of course. There may be positive/negative points about this.

**Comment by cristiandley** [comment]
https://github.com/apollographql/apollo-client/issues/6017

**Comment by cristiandley** [comment]
#6017

**Comment by 3nvi** [comment]
> I have witnessed exactly the same behaviour! It works fine when I refetch after I delete an item but not when I create one.

In case anyoone is still interested, this behavior happens because when you delete something, the  component holding the original query didn't get unmounted (normally because you just show a popup), while when creating a new item you redirect to another screen (i.e. create item page) unmounting the page holding the original query.

Apollo maintains a list of observable queries at any given time, which is based on which components are currently mounted that defined a query (i.e. through `<Query>...</Query>` or `useQuery`). When a component gets unmounted - at least in the React world - , Apollo removes it from the list of observable queries by calling `QueryManager.tearDownQuery` when the component unmounts. During `refetchQueries` , when given a Query name string, apollo searches only the observable queries and because it can't find it, it doesn't execute a refeetch. If you provide a `{ query: ..., variables: .. }` as a refetch value, Apollo bypasses this search and always executes the query you've given it, as if it was a completely new query (since it has the doc and the variables it needs to execute it)

I don't know if @hwillson would like to give us more context as to whether there is a way around this (except from not using the React or by making sure that we unmount as little as possible on crucial refetches).

In short:

**`refetchQueries` will only work with strings if the component that defined the original query is not unmounted. On the contrary, it will always work when using the  `{ query... ,  variables: ... }` style.**

**Comment by Emiliano-Bucci** [comment]
So it's not possible to perform a refetch with the original query and variables? Doesn't sound to me a good choice :/

**Comment by 3nvi** [comment]
> So it's not possible to perform a refetch with the original query and variables? Doesn't sound to me a good choice :/

It's only possible if you don't  navigate away (i.e. unmount the component that performed the  original query).

Your other option (which is not actually a solution) is to use a different `networkPolicy`  so that when you revisit the component that held the query, a server-query is performed.

**Comment by Emiliano-Bucci** [comment]
I see; thanks!

**Comment by Momepukku** [comment]
@3nvi 

> refetchQueries will only work with strings if the component that defined the original query is not unmounted.

I have three components render on the same page which use the same query(and same operation name) but difference variables.
when I specify operationName as refetchQueries, only the last one got refetch.
Why?

**Comment by 3nvi** [comment]
cause it literally does just that. It refetches the **query** with the latest variables. From apollo's side it's only 1 query with 3 different invocations, so i tries to refetch the "latest" invocation.

**Comment by slask** [comment]
Still an issue in 3.1.5 on May 22nd 2020

**Comment by moneebalalfi** [comment]
What happened with this issue, please!!

**Comment by Idan-Hen** [comment]
I think this needs to have option to refetch non observable queries.
and also not latest varibales - but all variables for this query 

those flags can solve a lot of issues

**Comment by jdmoliner** [comment]
Apollo client 3.2.5 and issue is not fixed. Any updates?

**Comment by andoks** [comment]
May be related to #3540

**Comment by andoks** [comment]
I am using Apollo client 3.2.9 and the issue is still there. I need to trigger refetch of a number of queries (for which has a pollInterval configured, if that has any implications in this context).

I tried working around the issue using fetchPolicy: no-cache, and also tried updating cache directly in the delete-mutation update option, neither worked.

However returning PureQueryOptions objects instead of the query name strings from refetchQueries worked.

**Comment by espinhogr** [comment]
I have a very similar problem to the one you are experiencing here, I'm on version 3.3.11. I migrated today from version 2.6.10 and this started happening, in apollo-client version 2 everything was working as expected.
I am basically using refetchQueries with a string array from a component A and I expect the component B to have the query refreshed when it's mounted again.
**But I've noticed something that hasn't been mentioned here.** If I run the code in development mode the code behaves as expected, if I do run in production mode (after the `yarn build`) I experience the same problem of refetchQueries not working. Not sure any of you experiences this as well.

**Comment by thomasjsk** [comment]
@espinhogr got the same issue. In my case refetch was performed on a component that was already unmounted, and that - for some reason worked well on dev, but didn't on prod. Check your warnings (red ones lol), maybe that's the case for you as well.

**Comment by mutefiRe** [comment]
> `refetchQueries` will only work with strings if the component that defined the original query is not unmounted. On the contrary, it will always work when using the `{ query... , variables: ... }` style.

on the other side, refetchQueries with the `{ query... , variables: ... }` style also refetches queries which never were queried before. Which is also a behaviour I would not prefer.

**Comment by manuFL** [comment]
Same here, working on dev but not on prod.... please help

**Comment by maxsalven** [comment]
This is still an issue on 3.4.7

Reproduction:
https://codesandbox.io/s/floral-sea-cg45v?file=/src/App.js

Navigate to 'Mutation' and click 'Mutate'.

You'll see a console warning:

> Unknown query named "Slim" requested in refetchQueries options.include array 

However as per the docs
https://www.apollographql.com/docs/react/data/mutations/#refetching-queries

> The name of a query you've previously executed, as a string

**Comment by kalote** [comment]
I faced this issue today.

Same situation (mutation from another component, not in the tree anymore, that need to refetch a query upon success).

Is there any workaround?

**Comment by Off2Race** [comment]
Hi, @kalote – I ran into this issue as well. The only workaround I could come up with was to use the new client.refetchQueries method, specify all "active" queries, and filter out the ones I didn't watch to refetch using the onQueryUpdated function. See the example below:

```
client.refetchQueries({
    include: 'active',
    onQueryUpdated(observableQuery) {
        switch (observableQuery.queryName) {
            case 'QueryIWantToRefresh':
                return true;
        }

        return false;
    },
});
```

**Comment by sepehrg** [comment]
I faced the same problem. Any updates?

**Comment by benjamn** [comment]
@Off2Race I would expect that code to be equivalent to the following:
```ts
client.refetchQueries({
  include: ["QueryIWantToRefresh"],
})
```
Does that not work for you? Or was your point just that you had to use `client.refetchQueries` instead of the `refetchQueries` option for mutations?

**Comment by Off2Race** [comment]
Hi, @benjamn – I actually haven't tried the simpler form you suggested but I think it might have the same problem. 

The real key to my workaround was making sure ```QueryIWantToRefresh``` is active _before_ it's refreshed. Using ```client.refetchQueries``` instead of the ```refetchQueries``` option was really just the means to do that. 

The ```onQueryUpdated``` option by itself is a great new tool for development. In this context, though, it was the only way I could figure out how to pre-validate that a query was active/mounted before asking that it be refreshed.

All that said, I can try the version you suggested. Would you expect it to work (i.e. not throw the error)?

**Comment by pbassut** [comment]
I can confirm @Off2Race solution works(the long version but not the short one from @benjamn. Despite them looking the same).
But that's is uhhhh-gly

**Comment by MainAdc95** [comment]
still, an issue for me too, am deleting some entities in one page and it's updating the cache as you would expect, but when I create an entity on another page it's doesn't update the cache when I come back to the entity listing page.

**Comment by benjamn** [comment]
Based on this discussion, I have a theory I'd like to test (see #8825), which is that `QueryIWantToRefresh` is not refetched because it happens to have no observers, even though the developer explicitly requested the refetch.

**Comment by Off2Race** [comment]
Hi, @benjamn – Once you have a new version with #8825, I'm happy to test with my use case. I'm on 3.4.8 now, so it should be an easy upgrade. Just let me know how I can help.

**Comment by Off2Race** [comment]
Hi, @benjamn – I just upgraded my project to 3.4.16 (which should have #8825) but I'm still getting the following warning when I use ```refetchQueries``` with mutations.

```
Unknown query named "QueryIWantToRefresh" requested in refetchQueries options.include array
```

In this case, ```QueryIWantToRefresh``` is used in various other components that weren't mounted at the time the mutation was executed. The warning appears benign (more a nuisance than anything). And fortunately, the workaround I posted earlier still seems to work. But I thought you'd want to know.

Is there other information I can share to help you figure out the cause? Should I log a new GitHub issue since this one is closed?

Thanks!

**Comment by szamanr** [comment]
having the same issue on `3.4.16`. the query from an unmounted component is not refetched, regardless of whether i specify it using `refetchQueries` when defining a mutation or call `apolloClient.refetchQueries`.

**Comment by daveslutzkin** [comment]
We have the identical issue to @szamanr.

@Off2Race - unfortunately it's not as simple as "When they do mount, the queries should update normally." Of course that does happen, but in this case "updating normally" is pulling the data from the cache, which is now out of date because of the mutation where we asked for refetch.

Partly this is a fundamental issue with the Apollo design, where instead of manual cache invalidation the choice was query refetching, which is a slightly different thing and harder to achieve (and provably more expensive). If you look at react-query (https://react-query.tanstack.com/guides/query-invalidation) it does it the other way and turns out to be much simpler and easier to use in practice.

@benjamn I had hoped that https://github.com/apollographql/apollo-client/pull/8825 would fix this but unfortunately I'm still seeing it in 3.4.16 post-merge of that PR.

Apollo needs to either:

* Flush the cache when queries stop being observed (correct but expensive).
* Work out a way to refetch instances of queries even though the variables for that instance have been discarded by QueryManager (presumably hard).
* Flush the cache when refetch is requested for a query which is no longer observed (ideal but I have no idea how hard that is).

**Comment by daveslutzkin** [comment]
What I'm doing to make this work is to just manually `evict`ing the data. Feels a bit more low-level than is ideal but I guess that's by design?

**Comment by szamanr** [comment]
@daveslutzkin i guess a workaround is to set the cache policy to "cache-and-network" on the request you're trying to refetch.

**Comment by giggo1604** [comment]
Looks like there is something wrong with the `defaultOptions`. As stated in the [docs](https://www.apollographql.com/docs/react/api/core/ApolloClient/#example-defaultoptions-object) the default fetch policy for a `watchQuery`, which is used by the `useQuery` hook, should be `cache-and-network`. If I manually set the `defaultOptions` on client creation everything works as expected.

**Comment by nilesh-05** [comment]
I am having the same issue. After a book gets inserted in my database, I need to get all books again to show them in the list. But refetchQueries is not working. 
`
const [addBook, response] = useMutation(CREATE_BOOK, {
    variables: { name, genre, authorID },
    refetchQueries: [{ GET_BOOKS }],
    awaitRefetchQueries: true,
});
`
Each time after adding a book, I have to manually refresh my page. 
Version: 
"@apollo/client": "^3.4.16",
"graphql": "^15.7.2",

**Comment by Luchanso** [comment]
I have simple workaround solution for this - just add that query in the same function with mutation:

```js
function ProductAdd() {
  const result = useQuery(GET_SHOP, { ... });
  const [...] = useMutation(
    CREATE_PRODUCT,
    {
      refetchQueries: ["GetShop"],
    }
  );
  // ...
}
```

**Comment by marcelchastain** [comment]
I was able to fix this issue in my local project, though the solution might not apply to everyone.

I'm using `getDataFromTree()` , then 
```typescript
const initialState = apolloClient.extract();
```

That's being passed to my page in the normal SSR fashion, via `pageContext` (or more commonly, `window.__APOLLO_STATE__` or something)

I found that in my app's page render/hydration function I needed to get ahold of that value and call
```typescript
apolloClient.restore(initialState)
```

My theory is that, in my particular setup, the `getDataFromTree()` caused the queries to be "helpfully" executed server-side, which is why I wasn't seeing the additional refetch network request in my browser as I was expecting.

If you're using anything similar, I'd say inspect your pageContext (or `window.__APOLLO_STATE__`) to see if the refetched data is indeed reflected there.

**Comment by Vuliniar** [comment]
It would be great if we could have support for refetching unmounted/unobservable queries.

**Comment by smithWEBtek** [comment]
I wonder why `refetchQueries` is working in this older example from Reed Barger's course:
https://github.com/smithWEBtek/refetch-queries/blob/cf039ac4328eecb3091e93a5f8b0277d8f959eb0/src/App.js#L73

**Comment by ddkang** [comment]
Why was this issue closed? It looks like there's a lot of errors even in recent releases? I'm also running into similar issues

**Comment by jpvajda** [comment]
Thanks everyone for the comments and feedback we appreciate it! The Apollo Client team is taking a look at this problem again and will keep you all informed of when a fix will be released.

**Comment by daveslutzkin** [comment]
Just to repeat my comment from above more concisely, the goal here as a developer is to be able to say "please no longer use the cached result of this query which I now know is invalid". Whatever mechanism you can come up with for simply achieving that would be hugely appreciated.

(I know it can be done by manual cache manipulation but that's way more painful than is desirable in most situations.)

**Comment by daveslutzkin** [comment]
@jpvajda Any update/thoughts on ergonomic improvements here? Obviously not magically fixed yet, that's fine, but would love to hear if there's progress.

**Comment by tiagosilvaalopes** [comment]
Same here, I have **component A** that calls a mutation that then should refetch a query on **component B**, both components are never unmounted since the mutation does not remove anything and with the array of strings on refetchqueries it doesn't seem to be doing anything

**Comment by jpvajda** [comment]
@daveslutzkin Thanks for being patient on this issue, we know it's a big problem for many. 😅  We were considering this issue for an upcoming release, but had to move some things around due our current focus on `useBackgroundQuery` and `useFragment` which are 2 new hooks schedule to be released in [3.7](https://github.com/apollographql/apollo-client/milestone/28) 

I definitely want to bring this back to the team to discuss and try to determine when we may be able to resolve it.  I'll keep everyone posted on this issue.

**Comment by efstathiosntonas** [comment]
Hi, any updates?

**Comment by jpvajda** [comment]
Hi @efstathiosntonas 👋  thanks for checking in on this issue. We are currently focused on several other releases ahead of resolving this issue, so I have no ETA on it's resolution at this time. Once we determine a release, will add this issue to a [Github Milestone ](https://github.com/apollographql/apollo-client/milestones)for that release which will indicate it's prioritized. We are do our best to keep these milestones fairly accurate, so the community has a sense of what is being prioritized on our side.

**Comment by zaytsev-tech** [comment]
Hi all! My team couldn't wait release with this fix, because I had to hack apollo client :)
Instead client.refetchQueries I used refetchQueries into mutation and set array [query: ... , variables: ...]. But how I get all DocumentNode of active queries?
I have implemented this hook:
`
export const useGetActiveQueriesFromApollo = () => {
  const client = useApolloClient() as any; //any to get access to private props

  const queryDocumentNodes = new Set();

  const initialQueries = client.queryManager.queries; //we get all queries

  for (const queryDocument of initialQueries.values()) {
    queryDocumentNodes.add({ query: queryDocument?.document, variables: queryDocument?.variables });
  }

  return Array.from(queryDocumentNodes);
};
`

By applying this hook, we get all queries maybe which are repeated in array. This is not good for optimisation. The solution would be to use uniqBy from lodash.

The decision for client.refetchQueries is controversial (we breaking typescript patterns of private fields) but it work for me.

**Comment by salemhilal** [comment]
I've had a similar problem, and I've tried @zaytsev-tech's approach as a fix. 
For context, I have two pages: a list page and a "new list item" page. I have a two queries; a query to get all the current list items and a mutation to create a new one. The mutation should refetch the query (among a few others). Here's what I've noticed by inspecting the queries in `client.queryManager.queries`:

1. When I navigate to the "new list item" page, the query for the list page disappears from the query manager's list of queries.
2. The cache doesn't seem to have a reference to this query either.
3. I do see references to a few queries that power the "new list item" page.
4. When I make a new list item and navigate back to the standard page (navigation here done with Next.js's router), the query reappears and the new queries from the "new list item" page vanish, almost as if each page has its own cache. 
5. As far as I can tell, I have one reference to the client, which stays constant throughout the whole flow so I don't think I'm creating multiple copies of the client.

Since I can't evict a query that doesn't seem to be in the cache in the first place, using `refetchQueries` doesn't work, with or without using `useGetActiveQueriesFromApollo` like above.

I'm going to dig a bit more, but I figured my data point would be worth sharing.

**Comment by salemhilal** [comment]
I have an update! I kept an eye on the Apollo dev tools, and here's what I noticed:

1. When I move between pages, queries get evicted and added as each component gets evicted and added. On the list page, I can see the list query. On the "new list item" page, I can see the new list item queries. This feels like what's expected.
2. However, the cache only ever grows. When I navigate to the list query page after making a new item, the cache appears to have all the data I need. My list page query, having been evicted, looks at the cache, sees the data it needs, and then uses that data (even though the list is now incomplete). 

I think the problem I'm seeing is that `refetchQueries` doesn't have a full picture of all the queries that have previously ran. For this reason, providing just the name of a previous query isn't enough, and providing a query from the query manager may not be enough either. 

I could manually write some cache eviction logic to handle this particular case, but that sounds cumbersome given that I know the query I want to refetch.

I'll note that I only see queries getting evicted from the query manager in production, and not in dev or local builds of my application. I'm not sure why that's happening. I can't tell if the query evictions are supposed to happen, or if it's an indication of a bug.

**Comment by salemhilal** [comment]
Ok, sorry for triple-posting here. I had a lot of time to kill on a train. 

**tl;dr:** `refetchQueries` has a history of not working with queries owned by unmounted components. There was a fix, but it was deleted.

Here's what I've figured out from picking over a bunch of Github issues (some of this may be wrong and I'm new to Apollo's innards, so take it with a grain of salt):

- `refetchQueries` and `updateQueries` both used to work like you’d think (give them a query that was run before and it'll run it again). It worked because Apollo Client (then `apollo-react`) inadvertently kept a reference to every query. 
- They stopped working because keeping queries around was expensive and would rerender components that were no longer mounted. The fix involved evicting queries that were no longer used by mounted components, but the fix in turn broke the nice behavior of `refetchQueries` and `updateQueries`. ([bug here](https://github.com/apollographql/apollo-client/issues/993), and [this PR has a better description of the problem](https://github.com/apollographql/react-apollo/pull/462)).
- To fix this, a "Recycler" was added. The idea was to let one copy of every query live forever. When a query was no longer needed actively, the recycler kept it around and let it receive updates from `updateQueries` without triggering re-renders. Since queries were kept around forever, you could reference them any time in the future in `refetchQueries` and `updateQueries`. Everyone rejoiced ([PR here](https://github.com/apollographql/react-apollo/pull/462)). 
- Alas, the Recycler turned out to be a good idea in theory, but in practice caused “a lot of loading pain” for people. It was removed [in this PR](https://github.com/apollographql/react-apollo/pull/1715), which also added a shim to allow `refetchQueries` to continue accepting an array of strings "temporarily". I'm not clear if or how the problem that the Recycler solved was fixed in another way. 
- Some point after that, `react-apollo` was merged into the `@apollo/client` repo. `refetchQueries` still accepts querying by name, but I'm not sure if that's on purpose or if it's just because this is an old problem in a big repo. I'm not sure if there's a catalog of all previously-called queries somewhere accessible from the client.

The solution I'm going to try is to use the `update` option [to evict the part of the cache that's stale](https://www.apollographql.com/docs/react/data/refetching/#refetching-queries-affected-by-cache-updates) and force any related queries to rerun. Maybe something like this: 

```ts
useMutation(AddListItemQueryDocument, {
    variables: { ... },
    update(cache) {
        // evict the stale part of the cache
        cache.evict({ fieldName: "myList" })
    }
});
```

**Comment by jpvajda** [comment]
@salemhilal  no worries! It's actually can be easier to parse comments in smaller chunks that one larger one. Thanks for sharing your experience here.

**Comment by dylanwulf** [comment]
> The solution I'm going to try is to use the update option [to evict the part of the cache that's stale](https://www.apollographql.com/docs/react/data/refetching/#refetching-queries-affected-by-cache-updates) and force any related queries to rerun. Maybe something like this:

I've tried this approach several times and the problem I keep running into is that evicting the cached data makes it disappear from my screen. I want the old data to continue being displayed while refetching, but I don't see an easy way to do that with `cache.evict()`

**Comment by salemhilal** [comment]
Just for the sake of sharing what I found, evicting data from the cache works like a charm. 

It's nice because I don't have to think about all possible queries that could be impacted by my mutation — I just specify the part of the cache to evict and the queries refetch themselves the next time they're active. 

Here's roughly what I have:
```ts
const [addListItem] = useMutation(AddListItemMutationDocument, {
    // Evict outdated information from our cache after running the mutation.
    update(cache: ApolloCache<NormalizedCacheObject>) {
        // Get the ID of the document in the cache we want to evict part of.
        const id = cache.identify({ __typename: "List", id: currentListId });
        // Evict the part of the document that is no longer correct.
        const fieldName = "listItems";
        cache.evict({ id, fieldName });

        // "Evicting an object often makes other cached objects unreachable.
        //    Because of this, you should call cache.gc after evicting one or
        //    more objects from the cache."
        // https://www.apollographql.com/docs/react/caching/garbage-collection#cacheevict
        cache.gc();
    },
});
```

If I wanted this code to be more efficient and a little more clever, it could manually update my cache locally with my new data rather than re-querying for everything. But as a first step, this seems fine for most of our use cases.

**Comment by alan345** [comment]
I am having the same issue: 
If I run the code in development mode the code behaves as expected, if I do run in production mode (after the yarn build) I experience the same problem of refetchQueries not working.

The first component has this query:
```
const QUERY = gql(`
  query perksTribeQueryConnection_PerksTribeQuery($orderBy: PerkOrderByInput, $where: PerkWhereInput, $skip: Int, $first: Int) {
    perksTribeQueryConnection(orderBy: $orderBy, where: $where, first: $first, skip: $skip) {
      perks {
        id
        isPublished
        partner {
          id
          name
        }
        perkDetail {
          id
          name
          offer
          details
          typeCreationPerkDetail
          estimatedSavingMax
          product {
            id
            nameFile
          }
        }
      }
    }
  }
`)

```

I run in another component.
```
      await client.refetchQueries({
        include: ['perksTribeQueryConnection_PerksTribeQuery'],
      })
```

It will trigger the refetch of the query `perksTribeQueryConnection_PerksTribeQuery` as expected in dev, but NOT in Prod.

**Comment by barbalex** [comment]
This issue has been open for 4 years. It is one of the reasons we are migrating to tanstack query.

**Comment by daveslutzkin** [comment]
I don't think apollo can fix this while preserving backwards compatibility, it's a fundamental design issue with the library unfortunately. Probably worth moving on past apollo at this point, it was great five years ago but just ok now.

**Comment by alessbell** [comment]
Hi all 👋 

I was trying to understand whether this is still an issue, so I forked a sandbox mentioned above and updated Apollo Client to the `latest` version, `3.7.11`. By visiting the `Query` page first, then navigating to the `Mutation` page and clicking "mutate" you can see both the mutation and the refetch (triggered via `refetchQueries: ["Slim"]` in `App.js`) fire: https://codesandbox.io/s/unruffled-napier-wzdv9m?file=/src/App.js:715-739

If anyone has any other questions here, do let me know! Given that I cannot reproduce the bug in the latest version, I'm going to close this issue out, but it will remain unlocked for any follow-up comments/questions for 30 days. Thanks!

**Comment by sakhmedbayev** [comment]
It is still an issue in `^3.7.11` version

**Comment by amaster507** [comment]
> Given that I cannot reproduce the bug in the latest version, I'm going to close this issue out

I just came across this myself in `^3.7.11`. So yes, it is still an issue.

**Comment by dizzyjaguar** [comment]
Running into this right now with a delete `modal` and a deleting an item from a `datagrid`. Was hoping to trigger a re-query for the `datagrid` but it's not recognizing my string query.

**Comment by jerelmiller** [comment]
@amaster507 interesting solution! Feel free to keep using it, but I'd recommend a couple tweaks that is more safe (i.e. not subject to breaking changes) and doesn't require you to reach into the internals as much.

1. Rather than getting the queries like this:

```ts
const initialQueries = client.queryManager.queries as Map<string, QueryInfo>;
```

Use `client.getObservableQueries('active')` instead. If you want all queries, not just the active ones, you can use `client.getObservableQueries('all')`. This will return a map of query IDs to `ObservableQuery`. Here you can get the query from the `query` property:

```ts
const activeQueries = client.getObservableQueries('active');
const queryDocuments = Array.from(activeQueries.values())
  .map(observableQuery => observableQuery.query);
```

`client.getObservableQueries` is a public method and won't be subject to breaking changes, so it would be safer to use this.

2. It looks like you don't use `variables` when getting the intersection of of query names, so you should be able to omit that from your `queryDocuments` type.

3. When mapping over your `queryDocuments`, your implementation assumes the query definition comes first with fragments defined afterward. Fragments can also be declared before the query operation and be considered a valid GraphQL document. If you'd like this to be a bit more robust, consider using the `getMainDefinition` function exported from `@apollo/client/utilities` (you can see an example [here](https://www.apollographql.com/docs/react/data/subscriptions#3-split-communication-by-operation-recommended)). This will return an `OperationDefinitionNode` if your document contains one, or the first `FragmentDefinitionNode` if your document only contains fragments. From here you can extract the name.


```ts
import { getMainDefinition } from '@apollo/client/utilities';
return intersect(Object.values(queryDocuments).map(doc => {
  const def = getMainDefinition(doc.query);

  if (def?.kind === 'OperationDefinition') {
    return def.name?.value ?? ''
  }
  return ''
}).filter(name => name !== ''), toRefetch)
```

---

@dizzyjaguar I'm curious, when you're seeing this issue, is the query you're trying to refetch an active query? It's possible its not refetched for that reason (in which case the helper by @amaster507 would help). If you'd like to refetch regardless of whether the query is active, consider passing the query document to your `refetchQueries` instead of the string:

```ts
// where MY_QUERY is a query wrapped in `gql`
import { MY_QUERY } from './some/path';

mutate({
  // ...
  refetchQueries: [MY_QUERY]
})
```

If the query is active and you're still not able to refetch using a string array, we could use some help providing a reproduction where we can look at this a bit more in-depth. As @alessbell tried [above](https://github.com/apollographql/apollo-client/issues/5419#issuecomment-1497807784) this seems to be functioning as we the maintainers expect, but there could be something else that we just aren't seeing. A reproduction goes a long way to help us better diagnose the issue.

Hope this helps!

**Comment by daveslutzkin** [comment]
@jerelmiller 

I said this earlier in the thread but I'll try to outline it again. I think the problem is one of ergonomics not functionality - as you say, "functioning as we the maintainers expect".

When using the library, generally people just want to say "I now know that this data isn't valid so refetch it if I ask for it again".

But what apollo-client expects of people is to say "I now know that this specific active query isn't valid so please refetch it now", which is a much less predictable thing and very hard for a developer to work with in the general case.

For instance, it's very possible that a mutation will invalidate data in a way that I can predict, but that the queries that reference that data aren't particularly predictable. Or if they are predictable now, they won't be as the structure of the app inevitably changes. I don't care if the queries are active or not, and in general I can't predict whether they will be, because app code is complicated and the same mutation can be accessed from multiple places.

In that case what apollo-client expects of me is to write manual cache manipulation, which is complex and error-prone code, and again can change unpredictably as the structure of the app and therefore the cache changes.

This is completely workable and yet pretty frustrating and expensive to work with. Ideally what apollo-client would have instead is some nice shorthand for "invalidate this piece of data when this mutation returns" and then I could just pass that and move on, rather than spending half an hour reverse-engineering the structure of the cache and writing specific code to invalidate the pieces that matter.

Does that make any sense?

**Comment by barbalex** [comment]
> Does that make any sense?

It makes more clear what I have felt for quite a while: apollo is way too complex for simple cases. I have never fully understood it and it seems I have a lot to learn.

Which is why I have moved on to react-query. Where cache invalidation is so simple, even I have grasped it.

**Comment by daveslutzkin** [comment]
@jerelmiller 

> That being said, there are some performance-related concerns that we are also trying to balance here and this is something I think we do a poor job of communicating if I'm being honest. When queries stop being used, we tend to throw them away and allow the garbage collector to do its job so that we can avoid memory leaks with lots of objects building up over time. Its difficult to know when a query might or might not be used again. Since queries can "disappear" when they aren't being used, it makes sense that a query that has been thrown away can no longer be found by its name, hence why you're seeing the warning about a missing query.

I think that's perfectly understandable. But my point here is that I don't actually want to request refetch of a query by name, I just do that because it's the way the library seems to suggest (by providing the `refetchQueries` argument in the first place and documenting it). I actually want to invalidate/evict a piece of data, and this would also completely avoid the very valid performance concerns you raise.

**Comment by jerelmiller** [comment]
@daveslutzkin ah gotcha, apologies that I misunderstood what you were getting at. 

Perhaps this is also something not well communicated in our docs? You can do this very thing with the [`cache.evict()`](https://www.apollographql.com/docs/react/caching/garbage-collection/#cacheevict) function. So if you wanted to remove the a `Book` with `id` of `5`, you could do something like this:

```ts
const book = { __typename: 'Book', id: 5, ...otherFields }

cache.evict({ id: cache.identify(book) })
```

And if you want to remove dangling references to the evicted record, you can call [`cache.gc()`](https://www.apollographql.com/docs/react/caching/garbage-collection/#cachegc)

Is this what you're looking for?

**Comment by daveslutzkin** [comment]
@jerelmiller Hm, maybe? Though that should presumably be taken care of transparently by returning an object with id in the mutation response, so maybe I'm confusing myself and describing it wrong - maybe it is a query (usually a list query) not a piece of data.

Just looking through our code, most of the `refetchQueries` arguments are either list or aggregate queries that we know are invalidated in a particular place.

So maybe my point is more that we don't care and don't want to care whether those queries are "active". That feels like the library's concern not ours, purely an optimisation detail.

OK, so maybe what we want is to be able to say "invalidate this query so next time its result is requested, it doesn't come from the cache because we now know that cached value will be invalid".

So then:

* If the query is currently being used by a mounted component, it refetches instantly.
* If the query isn't currently being used by a mounted component, but then is requested by a component which is mounted later, it doesn't come from the cache but is then refetched.
* If the query is never requested again, it's never refetched.

Which I think is what happens when you call `cache.evict({ fieldName: "queryName" })`? Correct me if I'm wrong!

But because:

a) there's no `evictQueries` option to mutations; and
b) `evict` isn't mentioned in the docs here https://www.apollographql.com/docs/react/data/mutations/#updating-the-cache-directly

then it doesn't feel like it's a recommended way to proceed.

**Comment by jerelmiller** [comment]
@daveslutzkin thanks for this clarification. I think you and I are probably using the same terms to mean different things, so I'm probably confusing you even more 😬. Apologies for that!

Something to clarify is that Apollo's `InMemoryCache` is [normalized](https://www.apollographql.com/docs/react/caching/overview#data-normalization), so the concept of "invalidating a query" isn't quite right. When I think of "invalidating a query", I typically couple that concept with whole query caching.

Rather than thinking of this as "invalidating a query", think of it more as "invalidating a field or entity" in the cache, where an "entity" is defined as a chunk of data typically keyed by its `__typename` and `id` (one might also call this a "model". `Book` and `User` might be an "entity" for example). Apollo's cache is built so that if it can't fulfill all of the data needs for a query from the cache, it will go and fetch that data from the server. This concept is what I think you're alluding to when you say this:

> invalidate this query so next time its result is requested, it doesn't come from the cache because we now know that cached value will be invalid

You probably already know this information, but I wanted to state this to ensure we are talking about the same thing. If you're curious about a more in-depth look at whole query caching vs normalized caching and their behaviors, let me know and I'd be happy to add an example to explain this.

> most of the `refetchQueries` arguments are either list or aggregate queries

This is super helpful to know! In this case, the `cache.evict()` suggestion I gave you is likely not going to work well since you're looking to invalidate a _field_ on a record, not evict an entire record entirely from the cache. 

> my point is more that we don't care and don't want to care whether those queries are "active". That feels like the library's concern not ours, purely an optimisation detail.

I totally hear you here and and agree the current implementation is a bit of a limiting detail and a [leaky abstraction](https://en.wikipedia.org/wiki/Leaky_abstraction). We'll see if we can come up with some strategies for making this operate a bit more reasonably while allowing for the optimizations to take place, but this is likely not something we will look at until after we release v3.8 (sometime in the next few weeks).

Again, I'd still like to at least determine if this is a bug even with "active" queries so that we can classify this issue accordingly. A fix for active queries not refetching would likely be a quicker fix than determining a new strategy to make this work for all previously called queries. If you have some more details to share here to help classify this behavior, that would be helpful.

> OK, so maybe what we want is to be able to say "invalidate this query so next time its result is requested, it doesn't come from the cache because we now know that cached value will be invalid".

If `refetchQueries` with a string array is not working well for you, here are a couple ways I can think of that should work now in the latest version of Apollo (`v3.7.14` at the time of this writing).

### Solutions

1. `refetchQueries` _does_ also take an item in the array that is an object with a `query` and `variables` property. This method of refetching queries will guarantee a refetch on the query, even if the query is inactive. 

```ts
refechQueries: [{ query: QUERY, variables: {...} }]
```

Take this approach with a grain of salt and a bit of caution if you decide to try this. Our codebase literally calls this approach ["legacy"](https://github.com/apollographql/apollo-client/blob/2ebccfcefe8950d1e9322e11273cfe3e97ea989c/src/core/QueryManager.ts#L798) and we don't document it. I don't have the history or context to fully understand why this is the case, so it may come with its own quirks.

2. Consider modifying the cache directly to invalidate the field that contains the list of data you'd like to refetch. This approach jives with the way you are thinking about this:

> invalidate this query so next time its result is requested, it doesn't come from the cache because we now know that cached value will be invalid

`useMutation` or `client.mutate` (not sure which API you're using) has an [`update` option](https://www.apollographql.com/docs/react/data/mutations#update) that allows you to make direct cache updates. Here you can invalidate the field that contains the list. The next time you go to execute the query with this field, Apollo will see that it can only fulfill partial data for the query and will go fetch data from the server for you.

You can use [`cache.modify`](https://www.apollographql.com/docs/react/caching/cache-interaction#using-cachemodify) to handle this behavior. We include an [`INVALIDATE` sentinel object](https://www.apollographql.com/docs/react/caching/cache-interaction#example-invalidating-fields-within-a-cached-object) in modifier functions that allows you to invalidate data for a field. This approach might look something like this:


```
mutate({
  update(cache) {
    // NOTE: You may have to provide an `id` option here if the field you're looking to
    // invalidate is not a top-level query field. 
    cache.modify({
      fields: {
        listField: (_, { INVALIDATE }) {
          return INVALIDATE;
        }
      }
    })
  }
});
```

Check out the list of [examples](https://www.apollographql.com/docs/react/caching/cache-interaction#examples) on how `cache.modify` might help you here.

Does this help a bit more?

**Comment by dylanwulf** [comment]
> We include an [INVALIDATE sentinel object](https://www.apollographql.com/docs/react/caching/cache-interaction#example-invalidating-fields-within-a-cached-object) in modifier functions that allows you to invalidate data for a field.

It is my understanding that `INVALIDATE` only causes queries to re-read that field from the cache, not refetch, am I misunderstanding?

**Comment by jerelmiller** [comment]
@dylanwulf oh you might be right. All this talk of invalidating data got my brain scrambled 😂. 

`DELETE` is probably the better option here as it deletes a field from an object entirely. My example above should read:


```ts
mutate({
  update(cache) {
    // NOTE: You may have to provide an `id` option here if the field you're looking to
    // invalidate is not a top-level query field. 
    cache.modify({
      fields: {
        listField(_, { DELETE }) {
          return DELETE;
        }
      }
    })
  }
});
```

**Comment by dylanwulf** [comment]
@jerelmiller Ah that makes more sense, thanks! I've tried that in the past and it does work for refetching, but my main problem with it is that it causes my query data to disappear from the screen until the refetch completes. Not the best user experience, so I'm still sticking to `refetchQueries` for now.

**Comment by jerelmiller** [comment]
@dylanwulf there are definitely tradeoffs to each approach. If you know your query is "active" (like it sounds you have), then `refetchQueries` will probably work better for you. In the case of this issue where a query is inactive or has been cleaned up, the cache deletion might work better to force the query to re-execute when it becomes active again. Definitely a tradeoff to consider!

**Comment by sakhmedbayev** [comment]
What I ended up doing is just placing all `refetch` functions for all previous queries in the react context as dictionary and using it from there in my app.

```
const {..., refetch} = useQuery(...)

setAllPrevQueries(
  {
     ...allPrevQueries,
     keyOfSomePrevQuery: refetch
  }
)

// and then
const {allPrevQuereis} = useAllPrevQueries()

const handleRefetchSomewhere = () => {
    allPrevQuereis[keyOfSomePrevQuery]()
}

```

**Comment by abarreir** [comment]
> Hi all 👋
> 
> I was trying to understand whether this is still an issue, so I forked a sandbox mentioned above and updated Apollo Client to the `latest` version, `3.7.11`. By visiting the `Query` page first, then navigating to the `Mutation` page and clicking "mutate" you can see both the mutation and the refetch (triggered via `refetchQueries: ["Slim"]` in `App.js`) fire: https://codesandbox.io/s/unruffled-napier-wzdv9m?file=/src/App.js:715-739
> 
> If anyone has any other questions here, do let me know! Given that I cannot reproduce the bug in the latest version, I'm going to close this issue out, but it will remain unlocked for any follow-up comments/questions for 30 days. Thanks!

@alessbell you were not able to reproduce the issue because your app is rendered under React `StrictMode`, which renders components twice when running the app in dev mode. This has a side effect with apollo : each query is run twice, consequently **each query is duplicated in the observed queries array**.

When unmounting a component, only one of each duplicated queries is removed from said array, refetch is thus successful because it matches the query name with the leaked query in the array. I've removed `StrictMode` from your exemple and am able to reproduce the issue : https://codesandbox.io/s/epic-darwin-2jf6td?file=/src/index.js

This difference of behaviour between dev and prod build is also very confusing (took me an afternoon of debugging to understand what was going on) so we could at least work on removing the leak under strict mode ? Or maybe what's in the works for  3.8 will fix that ?

**Comment by DiRover** [comment]
Hi for everyone, today I got stuck with the same issue. I have component A that has a list and a form for creating items for this list. There is also component B that only has a list from component A. In dev mode, everything works correctly, but in prod mode, the list in component B does not update automatically when creating a new item, only after page reload.

```
const [createFun] = useMutation(GQL, {
        refetchQueries: [LIST_GQL], // DocumentNode object parsed with gql
    });
```

**Comment by andreimatei** [comment]
Hi @jerelmiller! I've hit this `Unknown query [object Object] requested in refetchQueries options.include array` error even though I'm trying to refetch queries by "document", not by string. Is this surprising?

You've said:

> If you'd like to refetch regardless of whether the query is active, consider passing the query document to your refetchQueries instead of the string:

And yet, I seem to see the same behavior when I'm using the document (or at least I think I am).

This is my code:

```ts
export const GET_COLLECTIONS = gql(`
  query GetCollections($input: GetCollectionsInput!){
    getCollections(input: $input) {
      id
      name
    }
  }
`);

const [captureSnapshotsMutation] = useMutation(CAPTURE_SNAPSHOTS, {
  refetchQueries: [GET_COLLECTIONS],
  // The GET_COLLECTIONS query is not always active when this mutation runs,
  // so we need to also manually evict its results from the cache. We really
  // want this query to be re-executed the next time the snapshots page is
  // rendered.
  update(cache) {
    cache.evict({id: "ROOT_QUERY", fieldName: "getCollections"});
  },
});
```

This results in `Unknown query [object Object] requested in refetchQueries options.include array` printed when the `GET_COLLECTIONS` query is not active. Notice the `[object Object]` in there - that's presumably saying something interesting.
One thing to note is that I'm using `"@graphql-codegen/cli` to generate types from my `gql`. The type of `GET_COLLECTION` thus is `TypedDocumentNode<GetCollectionsQuery, Exact<{input: GetCollectionsInput}>>`.

Is all this surprising / indicative of a bug?

**Comment by Emiliano-Bucci** [comment]
@jpvajda No news right? :)

**Comment by emil14** [comment]
Any updates? That's a critical bug

**Comment by jerelmiller** [comment]
Hey @emil14 👋 

I'm not actually sure if there is a bug here. From what we've established so far, this seems to be an issue more about ergonomics of the API than an actual underlying bug. Please take a look at the comments from me and @phryneas above.

Are you encountering something different something different than whats already been mentioned?

**Comment by vijayth2-cerebras** [comment]
This worked for me after creating a new one:
```
refetchQueries: [{ query: ListMyOrganizationsDocument }]
```

**Comment by luap2703** [comment]
To my and I think many others understanding, the whole idea of refetchQueries is that I can invalidate/refetch any query from any mutation (although they don't know their respective existing) just with a smooth query name Document reference.

Now being forced to invalidate/refetch queries either by 1. specifically pointing towards them with their exact configuration, or by 2. manually jumping into the cache to modify stuff seems to be a huge design flaw -  I mean, how would the query know which pagination/filtering variables are currently used on eventually +10 other queries being inactive but turning active once a create/delete operation has been performed?). 

Now being forced to (due to 1. my mutation now knowing the inactive queries and to 2. minimize boilerplate cache-modification code) set all list queries in the application to "network-only" to have a smooth UI when delete or create operations occur does kinda remove the whole idea of a cache.


Is there no way to allow a flag on the refetchQueries array objects to say, "refetch this query if it is active immediately, if it is not active, invalidate the query in cache (i.e., somehow mark/invalidate its prop in the ROOT_QUERY) and refetch next time needed?
I could think of an additional options prop for mutations like `invalidateCache: [DocumentNode]` or an additional flag on the refetchQueries arr-objects, like `refetchQueries: { query: DocumentNode, invalidateInCache: true}`

It feels like there have been plenty of different threads now reporting this unexpected and broadly unwanted behavior.

Here: https://github.com/apollographql/apollo-client/issues/7878

Or here: https://github.com/apollographql/apollo-client/pull/11357

Or here: https://github.com/apollographql/apollo-client/issues/5422

Since again a lot of time has passed since this was discussed last time, would you mind to look at this once again and do you know if any upcoming changes address this [phryneas](https://github.com/phryneas) [jerelmiller](https://github.com/jerelmiller)?

**Comment by UPSxACE** [comment]
Since this is the default behavior, refetchQueries is useless in my opinion

**Comment by phryneas** [comment]
@luap2703 I agree that this function is generally confusing, but it's also very widely used - a change to it would be breaking. 
And adding more functionality to it will only make it more confusing.

We hear your frustration, and there's a good chance that we'll come up with a replacement with a different name in the future, but it's probably not going to be very soon, I'm sorry.

**Comment by luap2703** [comment]
Thank you @phryneas for replying!

Since there's no major change incoming, what would you recommend as "best practice" to overcome the current limitation? Manually adjusting the cash?

**Comment by phryneas** [comment]
Tons of nuance there, unfortunately - in the end, what works best for you. 
I would probably combine it with a manual cache invalidation if I read your case here right, but for anyone else reading this: please don't take this as blanket advice, but apply your situation accordingly!

**Comment by StasNemy** [comment]
I fixed it removing refetchQueries.

Problem is that this query wasn't called yet I guess

**Comment by ryancrunchi** [comment]
Same issue for me.
I have a component querying a list of items with query name `ListGet`.
On another page (where list of items is not mounted), I can create a new item (mutations and so on...).
After the new item is created I call `refetchQueries({ include: ['ListGet'] })` but it throws warning message `Unknown query named "ListGet" requested in refetchQueries options.include array`, despite the list component was mounted previously.

👉 This only happens in prod build. In dev build, the ListGet is refetched.

Tested with apollo-client versions 3.9.11 to 3.11.8

**Comment by kkx64** [comment]
### Possible solution with expected outcome

If anyone stumbles upon this, I had a specific use case where we would write queries like so:
```ts
const getXQuery = (param1: string) => gql`
query get_x_query {
   get_x_query(args: {param: "${param1}"}) {
      field1
   }
}
```

I wanted to invalidate the cache only if `param1` was a certain value (e.g. an ID of an object). After some digging this is what ended up working perfectly:

```ts
client.cache.modify({
      fields: {
        'get_x_query': (existing, { storeFieldName }) => {
            if (storeFieldName.includes('the-id-I-want-to-invalidate-for')) {
              return undefined;
            }
          return existing;
        },
      },
    });
graphqlClient.cache.gc();
```

Then I wrote this helper function:

```ts
/**
 * Invalidate queries in the cache based on the query name and the string that the query contains.
 * @param queries - An array of objects containing the query name and the string to invalidate the query if it contains.
 * @param queries[].query - Which query to invalidate.
 * @param queries[].invalidateIfContains - Invalidate the query if it contains this string. Otherwise, invalidate the entire query.
 */
export const invalidateQueries = (
  queries: { query: string; invalidateIfContains: string }[],
) => {
  queries.forEach(({ query, invalidateIfContains }) => {
    graphqlClient.cache.modify({
      fields: {
        [query]: (existing, { storeFieldName }) => {
          if (invalidateIfContains) {
            if (storeFieldName.includes(invalidateIfContains)) {
              return undefined;
            }
            return existing;
          }
          return undefined;
        },
      },
    });
    graphqlClient.cache.gc();
  });
};
```

Which you can then use like so:

```ts
          invalidateQueries([
            {
              query: 'get_x_query',
              invalidateIfContains: `the-id-I-want-to-invalidate-for`,
            },
          ]);
```

Hope this helps someone out, I can see it also working even if you're not using functions for defining the query strings like in my project.

**Comment by benderillo** [comment]
> ### Possible solution with expected outcome
> If anyone stumbles upon this, I had a specific use case where we would write queries like so:
> 
> ```ts
> const getXQuery = (param1: string) => gql`
> query get_x_query {
>    get_x_query(args: {param: "${param1}"}) {
>       field1
>    }
> }`
> ```
> 
> I wanted to invalidate the cache only if `param1` was a certain value (e.g. an ID of an object). After some digging this is what ended up working perfectly:
> 
> ```ts
> client.cache.modify({
>       fields: {
>         'get_x_query': (existing, { storeFieldName }) => {
>             if (storeFieldName.includes('the-id-I-want-to-invalidate-for')) {
>               return undefined;
>             }
>           return existing;
>         },
>       },
>     });
> client.cache.gc();
> ```
> 
> Then I wrote this helper function:
> 
> ```ts
> /**
>  * Invalidate queries in the cache based on the query name and the string that the query contains.
>  * @param queries - An array of objects containing the query name and the string to invalidate the query if it contains.
>  * @param queries[].query - Which query to invalidate.
>  * @param queries[].invalidateIfContains - Invalidate the query if it contains this string. Otherwise, invalidate the entire query.
>  */
> export const invalidateQueries = (
>   queries: { query: string; invalidateIfContains: string }[],
> ) => {
>   queries.forEach(({ query, invalidateIfContains }) => {
>     graphqlClient.cache.modify({
>       fields: {
>         [query]: (existing, { storeFieldName }) => {
>           if (invalidateIfContains) {
>             if (storeFieldName.includes(invalidateIfContains)) {
>               return undefined;
>             }
>             return existing;
>           }
>           return undefined;
>         },
>       },
>     });
>     graphqlClient.cache.gc();
>   });
> };
> ```
> 
> Which you can then use like so:
> 
> ```ts
>           invalidateQueries([
>             {
>               query: 'get_x_query',
>               invalidateIfContains: `the-id-I-want-to-invalidate-for`,
>             },
>           ]);
> ```
> 
> Hope this helps someone out, I can see it also working even if you're not using functions for defining the query strings like in my project.

Thanks! This, perhaps, is the most suitable solution for many who posted here.
Should be mentioned though that one shall pass correct name as "query" for this to work. E.g.:

```typescript
const SELECT_STUFF = gql`
query GetMeStuff($name: String!) {
   stuffz(filters: {name: $name}) {
      id
      essence
   }
}
`;

```
So pass the `stuffz` as query name as that is what is in the cache.

**Comment by charpeni** [comment]
Hey 👋 

I know this is a long-running issue and that multiple issues have been brought here over the years, but I'm convinced that some of the issues described here could be related to an issue (https://github.com/apollographql/apollo-client/issues/12164) that just got fixed (https://github.com/apollographql/apollo-client/pull/12236) and released as [`3.12.4`](https://github.com/apollographql/apollo-client/blob/main/CHANGELOG.md#3124).

A quick summary of the issue is that when passing `DocumentNode` to `refetchQueries`, it used to be compared by reference, where identical `DocumentNode` could trigger an `Unknown query requested in options.refetchQueries` because their reference differs, even though they are technically identical and definitely an active query. I couldn't pinpoint the exact reason why it was happening (multiple references), but I believe one contributing factor, especially in production, could be bundle splitting, where a query document node could be duplicated or even instantiated multiple times.

I invite you to read the issue and the pull request if you would like to learn more about this issue. Otherwise, make sure to update to `3.12.4` and test it out! In our case, it fixed all weird behavior, where `refetchQueries` was only working half the time.
